private/bufpkg/bufmodule/bufmodulebuild/module_file_set_builder.go

```diff
@@ -56,56 +60,153 @@ func (m *moduleFileSetBuilder) build(
-		// The ModuleFileSet expects a Module, and its dependency Modules, but it is not OK for
-		// a Module to both be the input Module and a dependency Module. This means we need to check
-		// each module in the workspace to see if it is the same as the input module, and only add
-		// it as a dependency if it's not the same as the input module. We determine this by comparing
-		// each workspace module's WorkspaceDirectory with the input module's WorkspaceDirectory,
-		// where having the same WorkspaceDirectory means being the same module. This is correct
-		// because each module in a workspace are constructed with its WorkspaceDirectory set to its
-		// path relative to buf.work.yaml and this value is guaranteed to be unique within the same
-		// workspace. This is predicated on the input module belonging to the workspace, and it would
-		// be a bug if the input module doesn't belong to this workspace.
-		//
-		// We could also determine which modules could be omitted here, but it would incur
-		// the cost of parsing the target files and detecting exactly which imports are
-		// used. We already get this for free in Image construction, so it's simplest and
-		// most efficient to bundle all of the modules together like so.
-		for _, potentialDependencyModule := range workspace.GetModules() {
-			if moduleWorkspaceDirectory != potentialDependencyModule.WorkspaceDirectory() {
-				dependencyModules = append(dependencyModules, potentialDependencyModule)
+	if workspace == nil {
+		// If we don't have a workspace, we can simply include the module and its dependencies.
+		var dependencyModules []bufmodule.Module
+		for _, dependencyModulePin := range module.DependencyModulePins() {
+			dependencyModule, err := m.moduleReader.GetModule(ctx, dependencyModulePin)
+			if err != nil {
+				return nil, err
 			}
+			dependencyModules = append(dependencyModules, dependencyModule)
 		}
+		return bufmodule.NewModuleFileSet(module, dependencyModules), nil
+	}
+	// If we have a workspace, we need to merge its dependencies with other modules in the
+	// workspace, including transitive dependencies, and resolve conflicts.
+	dependencyModules, err := m.getDependenciesForWorkspaceModule(ctx, module, workspace)
+	if err != nil {
+		return nil, err
 	}
-	// We know these are unique by remote, owner, repository and
-	// contain all transitive dependencies.
+	return bufmodule.NewModuleFileSet(module, dependencyModules), nil
+}
+
+func (m *moduleFileSetBuilder) getDependenciesForWorkspaceModule(
+	ctx context.Context,
+	module bufmodule.Module,
+	workspace bufmodule.Workspace,
+) ([]bufmodule.Module, error) {
+	// We do a first pass to collect all module pins that we are going to resolve
+	// through the module reader. If there are any conflicts, we will need to go
+	// through dependency resolution.
+	var (
+		dependencyModulePins              []bufmoduleref.ModulePin
+		seenDependencyModulePinIdentities = make(map[string]bufmoduleref.ModulePin)
```

Comment:
nit: perhaps something like `seenDependencyIdentityStringToPin` to indicate it's a map?